gpflow/training/monitor/monitor.py

```diff
@@ -0,0 +1,790 @@
+
+
+class Monitor(object):
+    """
+    Main monitor class.
+    It's main purpose is to provide a callback function that will be hooked onto an optimiser.
+
+    In its initialisation it will create a MonitorContext object that will be passed to monitoring
+    tasks.
+    """
+
+    def __init__(self, monitor_tasks: Iterator[MonitorTask], tf_session: Optional[tf.Session]=None,
+                 global_step_tensor: Optional[tf.Variable]=None) -> None:
+        """
+        :param monitor_tasks: A collection of monitoring tasks to run. The tasks will be called in
+        the same order they are specified here.
+        :param tf_session: Tensorflow session the optimiser is running in.
+        :param global_step_tensor: the Tensorflow 'global_step' variable
+        (see notes in MonitorContext.global_step_tensor)
+        """
+
+        self._monitor_tasks = list(monitor_tasks)
+        self._context = MonitorContext()
+        self._context.tf_session = tf_session
+        self._context.global_step_tensor = global_step_tensor
+
+        self._start_timestamp = get_hr_time()
+        self._last_timestamp = self._start_timestamp
+
+    def __call__(self, *args, **kwargs) -> None:
+        """
+        Class as a function implementation. Normally it would be called by an optimiser when it
+        completes an iteration. The optimiser may pass some arguments, e.g. the state of trainable
+        variables. This arguments will be passed to all monitoring tasks in case they make sense
+        to some of them.
+        """
+        self._on_iteration(*args, **kwargs)
+
+    def start_monitoring(self) -> None:
+        """
+        This function should be called before starting the optimiser. It evaluates the global_step
+        variable in order to get its initial value. It also resets the starting timer since the
+        time set in the __init__ may no longer be accurate.
+        """
+        self._context.init_global_step = self._context.global_step
+        self._start_timestamp = get_hr_time()
+        self._last_timestamp = self._start_timestamp
+
+    def stop_monitoring(self, *args, **kwargs) -> None:
+        """
+        This function should be called when the optimisation is done. It sets the optimisation
+        completed flag in the monitoring context and runs the tasks once more. All extra arguments
+        will be passed to the monitoring tasks.
+        """
+        self._context.optimisation_finished = True
+        self._on_iteration(*args, **kwargs)
+
+    def print_summary(self, monitor_tasks: Optional[Iterator['MonitorTask']]=None) -> None:
+        """
+        Prints the tasks' timing summary.
+        :param monitor_tasks: List of monitoring tasks to print the summary for, defaults to
+        all tasks.
+        """
+        print("Tasks execution time summary:")
+        for mon_task in (monitor_tasks or self._monitor_tasks):
+            print("%s:\t%.4f (sec)" % (mon_task.task_name, mon_task.total_time))
```

Comment:
Major: print => logging

Reply:
The intention was to print into stdout, not log.

Reply:
I agree with @ahsimb. It's not really debug, it's really about user interaction.